extensions/guacamole-auth-restrict/src/main/java/org/apache/guacamole/auth/restrict/connectiongroup/RestrictConnectionGroup.java

```diff
@@ -0,0 +1,183 @@
+     * The name of the attribute that contains a list of hosts from which
+     * a user may not access this connection group. The presence of this
+     * attribute, absent the guac-restrict-hosts-allowed attribute, will allow
+     * access from all hosts except the ones listed in this attribute. The
+     * presence of this attribute coupled with the guac-restrict-hosts-allowed
+     * attribute will block access from any hosts in this list, overriding any
+     * that may be allowed.
+     */
+    public static final String RESTRICT_HOSTS_DENIED_ATTRIBUTE_NAME = "guac-restrict-hosts-denied";
+    
+    /**
+     * The list of all connection group attributes provided by this
+     * ConnectionGroup implementation.
+     */
+    public static final List<String> RESTRICT_CONNECTIONGROUP_ATTRIBUTES = Arrays.asList(
+            RESTRICT_TIME_ALLOWED_ATTRIBUTE_NAME,
+            RESTRICT_TIME_DENIED_ATTRIBUTE_NAME,
+            RESTRICT_HOSTS_ALLOWED_ATTRIBUTE_NAME,
+            RESTRICT_HOSTS_DENIED_ATTRIBUTE_NAME
+    );
+    
+    /**
+     * The form containing the list of fields for the attributes provided
+     * by this module.
+     */
+    public static final Form RESTRICT_CONNECTIONGROUP_FORM = new Form("restrict-login-form",
+            Arrays.asList(
+                    new TimeRestrictionField(RESTRICT_TIME_ALLOWED_ATTRIBUTE_NAME),
+                    new TimeRestrictionField(RESTRICT_TIME_DENIED_ATTRIBUTE_NAME),
+                    new HostRestrictionField(RESTRICT_HOSTS_ALLOWED_ATTRIBUTE_NAME),
+                    new HostRestrictionField(RESTRICT_HOSTS_DENIED_ATTRIBUTE_NAME)
+            )
+    );
+    
+    /**
+     * The remote address from which the user accessing this connection logged in.
+     */
+    private final String remoteAddress;
+    
+    /**
+     * Wraps the given Connection object, providing capability of further
+     * restricting connection access beyond the default access control provided
+     * by other modules.
+     *
+     * @param connectionGroup
+     *     The ConnectionGroup object to wrap.
+     * 
+     * @param remoteAddress
+     *     The remote address from which the user accessing this connection
+     *     logged in.
+     */
+    public RestrictConnectionGroup(ConnectionGroup connectionGroup, String remoteAddress) {
+        super(connectionGroup);
+        this.remoteAddress = remoteAddress;
+    }
+    
+    /**
+     * Returns the original ConnectionGroup object wrapped by this
+     * RestrictConnectionGroup.
+     *
+     * @return
+     *     The wrapped ConnectionGroup object.
+     */
+    public ConnectionGroup getUndecorated() {
+        return getDelegateConnectionGroup();
+    }
+    
+    @Override
+    public Map<String, String> getAttributes() {
+
+        // Create independent, mutable copy of attributes
+        Map<String, String> attributes = new HashMap<>(super.getAttributes());
+        
+        for (String attribute : RESTRICT_CONNECTIONGROUP_ATTRIBUTES) {
```

Comment:
A quick comment explaining what this loop is for / is doing would be handy - same for the very similar loop in `setAttributes()`.

Reply:
Added via rebase.